src/mlpack/core/data/string_encoding_dictionary.hpp

```diff
@@ -0,0 +1,334 @@
+/**
+ * @file string_encoding_dictionary.hpp
+ * @author Jeffin Sam
+ *
+ * Definition of the StringEncodingDictionary class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_STRING_ENCODING_DICTIONARY_HPP
+#define MLPACK_CORE_DATA_STRING_ENCODING_DICTIONARY_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/boost_backport/boost_backport_string_view.hpp>
+#include <unordered_map>
+#include <deque>
+#include <array>
+
+namespace mlpack {
+namespace data {
+
+/*
+ * Definition of the StringEncodingDictionary class.
+ *
+ * @tparam Token The type of the token that the dictionary stores.
+ */
+template<typename Token>
+class StringEncodingDictionary
+{
+ public:
+  //! A convenient alias for the internal type of the map.
+  using MapType = std::unordered_map<Token, size_t>;
+
+  //! The type of the token that the dictionary stores.
+  using TokenType = Token;
+
+  /**
+   * The function returns true if the dictionary caontains the given token.
+   *
+   * @param token The given token.
+   */
+  bool HasToken(const Token& token) const
+  {
+    return mapping.find(token) != mapping.end();
+  }
+
+  /**
+   * The function adds the given token to the dictionary and assigns a label
+   * to the token.
+   *
+   * @param token The given token.
+   */
+  template<typename T>
+  void AddToken(T&& token)
+  {
+    size_t size = mapping.size();
```

Comment:
I think there is no need to create an alias here, if it's only used once?

Reply:
The main reason is the following: I am not quite sure which part of the equality is performed first.
```c++
std::map<int, int> m;

/* m[0] adds (0, 0) to the dictionary and increases the size of the dictionary.
 * The result of the expression depends on the moment 
 * of the execution of m.size().
 */
m[0] = m.size();
```